qiime2/sdk/context.py

```diff
@@ -36,13 +47,62 @@ def get_action(self, plugin: str, action: str):
         try:
             action_obj = plugin_obj.actions[action]
         except KeyError:
-            raise ValueError("An action named %r was not found for plugin %r"
-                             % (action, plugin))
-        # This factory will create new Contexts with this context as their
-        # parent. This allows scope cleanup to happen recursively.
-        # A factory is necessary so that independent applications of the
-        # returned callable recieve their own Context objects.
-        return action_obj._bind(lambda: Context(parent=self))
+            raise ValueError(
+                "An action named %r was not found for plugin %r"
+                % (action, plugin))
+
+        # We return this callable which determines whether to return cached
+        # results or to run the action requested.
+        def deferred_action(*args, **kwargs):
+            # If we have a named_pool, we need to check for cached results that
+            # we can reuse
+            if self.cache.named_pool is not None:
+                arguments = {}
+                hashable_arguments = []
+
+                # Make args look the same as kwargs then process them together.
+                # Transform the args then update with kwargs to retain the
+                # order they were passed in
+                for k, v in \
+                        zip(action_obj.signature.signature_order.keys(), args):
+                    arguments[k] = v
+
+                arguments.update(kwargs)
+                for k, v in arguments.items():
+                    if isinstance(v, qiime2.sdk.Artifact):
+                        v = str(v.uuid)
+                    if isinstance(v, Metadata):
+                        _, fp = tempfile.mkstemp()
+                        v.save(fp)
+                        v = md5sum(fp)
+                        os.remove(fp)
+
+                    hashable_arguments.append({k: v})
+
+                hashable_arguments = make_hashable(hashable_arguments)
+                invocation = IndexedInvocation(
+                    plugin_action, hashable_arguments)
+
+                if invocation in self.cache.named_pool.index:
+                    outputs = {}
+                    for key in action_obj.signature.outputs.keys():
+                        output = self.cache.named_pool.index[invocation][key]
+                        outputs[key] = \
+                            qiime2.sdk.Result.load(self.cache.data / output)
```

Comment:
This should be `self.cache.named_pool.load()` right? That should handle collections correctly